plutus: 1.9 -> 1.15.0.1, cardano-api: 8.20 -> 8.29.1

locallycompact opened the pull request.

locallycompact committed:
plutus: 1.9 -> 1.15.0.1
cardano-api: 8.20 -> 8.29.1

ch1bo commented on hydra-cardano-api/src/Hydra/Cardano/Api/MultiAssetSupportedInEra.hs:
```diff
@@ -5,13 +5,13 @@ import Hydra.Cardano.Api.Prelude
 -- manipulating values that do not commit to a particular era.
 class HasMultiAsset era where
-  multiAssetSupportedInEra :: MultiAssetSupportedInEra era
+  multiAssetSupportedInEra :: MaryEraOnwards era
```
This type class has not the exact same purpose anymore. It is now to "witness" `MaryEraOnwards` for any `era` that includes those features.

Should: rename type class and method and update haddock

ch1bo replied:
Fixed in https://github.com/input-output-hk/hydra/commit/64530957d211e351d0f01be6fbdfa161e8f41c78

ch1bo commented on hydra-cardano-api/src/Hydra/Cardano/Api/ReferenceTxInsScriptsInlineDatumsSupportedInEra.hs:
```diff
@@ -5,7 +5,7 @@ import Hydra.Cardano.Api.Prelude
 -- write functions using inline datums that do not commit to a particular era.
 class HasInlineDatums era where
-  inlineDatumsSupportedInEra :: ReferenceTxInsScriptsInlineDatumsSupportedInEra era
+  inlineDatumsSupportedInEra :: BabbageEraOnwards era
```
This also changed semantics due to the way the upstream witness type is named and used.

Should: rename type class, method and update haddock

ch1bo replied:
Fixed in https://github.com/input-output-hk/hydra/commit/3bb28f1c11ce07f18f0f716afa18937b1cb52530

ch1bo commented on hydra-cardano-api/src/Hydra/Cardano/Api/ScriptDataSupportedInEra.hs:
```diff
@@ -5,10 +5,10 @@ import Hydra.Cardano.Api.Prelude
 -- manipulating scripts that do not commit to a particular era.
 class HasScriptData era where
-  scriptDataSupportedInEra :: ScriptDataSupportedInEra era
+  scriptDataSupportedInEra :: AlonzoEraOnwards era
```
This also changed semantics due to the way the upstream witness type is named and used.

Should: rename type class, method and update haddock

ch1bo replied:
Fixed in https://github.com/input-output-hk/hydra/commit/9b30463c929989deb2f3c1e78d1cd4699b65efe2

ch1bo commented on hydra-cardano-api/src/Hydra/Cardano/Api/Value.hs:
```diff
@@ -80,11 +78,6 @@ fromLedgerMultiAsset :: Ledger.MultiAsset StandardCrypto -> Value
-toLedgerValue :: Value -> Ledger.MaryValue StandardCrypto
-toLedgerValue =
-  toMaryValue
-
```
IIRC we had added these "alias" functions to have consistent naming.

Should: not remove it to keep consistency.

ch1bo commented on hydra-node/src/Hydra/Ledger/Cardano.hs:
```diff
@@ -144,8 +144,8 @@ mkSimpleTx ::
-  body <- createAndValidateTransactionBody bodyContent
-  let witnesses = [makeShelleyKeyWitness body (WitnessPaymentKey sk)]
+  body <- createAndValidateTransactionBody BabbageEra bodyContent
+  let witnesses = [makeShelleyKeyWitness ShelleyBasedEraBabbage body (WitnessPaymentKey sk)]
```
Commenting once, but applies to multiple places:

Avoid mentioning `Babbage` in packages outside of `hydra-cardano-api`. Currently, that package serves as an "anti-corruption" layer and hard-codes the era we target. Even if we never make it configurable, we want to limit the amount of code that needs changing. That's also the reason why we have all those patterns and alias functions in `Hydra.Cardano.Api`, which remove the need for call sites to pick an era.

Must: try to avoid "witness" values outside of `hydra-cardano-api`.

Concretely telling from this call site, I would propose to provide a fixed-era `createAndValidateTransactionBody :: TxBodyContent Era -> Either TxBodyError (TxBody Era)` in `hydra-cardano-api` and the same for `makeShelleyKeyWitness` and/or using type class constraints. Note the `Tx` return value of `mkSimpleTx` is a hard-coded `Tx Era = Tx BabbageEra` right now.

ch1bo reviewed (changes requested):
Let's try to only tie `hydra-cardano-api` to `cardano-api` and insulate the remaining code base as much as possible. See my comments for some ideas.

locallycompact committed:
hydra-cardano-api: reexport createAndValidateTransactionBody

locallycompact committed:
hydra-cardano-api: reexport makeShelleyKeyWitness

locallycompact committed:
Regenerate plutus scripts

ch1bo committed:
Acknowledge different canonical serialization of some auxiliary metadata

It seems like ledger changed some metada encoding between versions. We
had some specific auxiliaryData re-encoding fail, while most of the
golden data was roundtripping just fine. We suspect a slight difference
in the "canonical" serialization produced by the ledger **iff** a data
type was not serialized before. Which is the case in our golden test
suite where we generated random values and serialize them.

TLDR; our golden tests are quite strict and this should be fine

ghost commented:
Seems like there is some "funny" test errors which are hard to troubleshoot: https://github.com/input-output-hk/hydra/actions/runs/7045153258/job/19174290641?pr=1168#step:5:1856

locallycompact committed:
hydra-plutus: split getHeadAddress into getHeadInput and getHeadAddress

locallycompact committed:
getHeadInput: Use case split instead of fromMaybe

locallycompact committed:
Regenerate plutus scripts

locallycompact committed:
nix flake update

locallycompact committed:
LogFilterSpec: update txids

ch1bo commented on cabal.project:
```diff
@@ -12,8 +12,8 @@ repository cardano-haskell-packages
-  , hackage.haskell.org 2023-10-20T06:54:18Z
-  , cardano-haskell-packages 2023-10-20T16:16:59Z
+  , hackage.haskell.org 2023-12-27T06:54:18Z
+  , cardano-haskell-packages 2023-12-27T16:16:59Z
```
Should: use actual index dates to remove warnings printed

ch1bo commented on hydra-cluster/datasets/1-node.cbor:
These changes should not be needed (that was the point of doing CBOR in #1195)

Must: remove this commit and check whether this is now working without re-generating.

locallycompact replied:
You do need to regenerate them.

```
lc@earthrealmo ~/S/g/i/h/hydra-cluster (lc/plutus-1.15)> nix develop .#benchs.hydra-cluster
 --command bench-e2e datasets datasets/3-nodes.cbor datasets/1-node.cbor --output-directory $(pwd)/../benchmarks --timeout 1000s
path '/home/lc/Source/github.com/input-output-hk/hydra/hydra-cluster' does not contain a 'flake.nix', searching up
Reading dataset from: datasets/3-nodes.cbor
Reading dataset from: datasets/1-node.cbor
Test logs available in: /tmp/nix-shell.0gcAgd/bench-bbb4050f1a39ec5b/1/test.log
Starting benchmark
Seeding network
Initializing Head
Comitting initialUTxO from dataset
HeadIsOpen
Client 2 (node 1): 0/3000 (0.00%)
Client 3 (node 2): 0/3000 (0.00%)
Client 1 (node 0): 1/3000 (0.03%)
Client 2 (node 1): 1/3000 (0.03%)
Client 3 (node 2): 1/3000 (0.03%)
Client 1 (node 0): 1/3000 (0.03%)
Client 2 (node 1): 1/3000 (0.03%)
Client 3 (node 2): 1/3000 (0.03%)
Client 1 (node 0): 1/3000 (0.03%)
Client 2 (node 1): 1/3000 (0.03%)
Client 3 (node 2): 1/3000 (0.03%)
Client 1 (node 0): 1/3000 (0.03%)
Client 2 (node 1): 1/3000 (0.03%)
Client 3 (node 2): 1/3000 (0.03%)
Client 1 (node 0): 1/3000 (0.03%)
Logfile written to: /tmp/nix-shell.0gcAgd/bench-bbb4050f1a39ec5b/1/logs/hydra-node-2.log
Logfile written to: /tmp/nix-shell.0gcAgd/bench-bbb4050f1a39ec5b/1/logs/hydra-node-1.log
Logfile written to: /tmp/nix-shell.0gcAgd/bench-bbb4050f1a39ec5b/1/logs/hydra-node-0.log
Logfile written to: /tmp/nix-shell.0gcAgd/bench-bbb4050f1a39ec5b/1/logs/cardano-node.log
Test logs available in: /tmp/nix-shell.0gcAgd/bench-bbb4050f1a39ec5b/2/test.log
Starting benchmark
Seeding network
Initializing Head
Comitting initialUTxO from dataset
HeadIsOpen
Client 1 (node 0): 0/3000 (0.00%)
Client 1 (node 0): 1/3000 (0.03%)
Client 1 (node 0): 1/3000 (0.03%)
Client 1 (node 0): 1/3000 (0.03%)
Client 1 (node 0): 1/3000 (0.03%)
Logfile written to: /tmp/nix-shell.0gcAgd/bench-bbb4050f1a39ec5b/2/logs/hydra-node-0.log
Logfile written to: /tmp/nix-shell.0gcAgd/bench-bbb4050f1a39ec5b/2/logs/cardano-node.log
Benchmark failed at bench/Bench/EndToEnd.hs:389:33 in main:Bench.EndToEnd: waitMatch did not match a message within 20s
  nodeId:           0
  seen messages:

To re-run with same dataset, pass '--work-directory=/tmp/nix-shell.0gcAgd/bench-bbb4050f1a39ec5b/1' to the executable
Benchmark failed at bench/Bench/EndToEnd.hs:389:33 in main:Bench.EndToEnd: waitMatch did not match a message within 20s
  nodeId:           0
  seen messages:

To re-run with same dataset, pass '--work-directory=/tmp/nix-shell.0gcAgd/bench-bbb4050f1a39ec5b/2' to the executable
lc@earthrealmo ~/S/g/i/h/hydra-cluster (lc/plutus-1.15) [1]>
```

ch1bo commented on hydra-cluster/test/Test/LogFilterSpec.hs:
This is the same issue as it was with the benchmarks. The fact that we are persisting JSON serialized transactions here which can change their byte representation (which hashed gives the transaction `id`) is fragile.

No need to change something in this PR, but this is yet another argument against having JSON of transactions and maybe only allow CBOR encoding of them (+ prepare / select tooling to easily introspect transaction CBOR)

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Wallet.hs:
```diff
@@ -176,7 +177,7 @@ newTinyWallet tracer networkId (vk, sk) queryWalletInfo queryEpochInfo = do
     makeShelleyAddress networkId (PaymentCredentialByKey $ verificationKeyHash vk) NoStakeAddress
 
-  ledgerAddress = toLedgerAddr $ shelleyAddressInEra @Api.Era address
+  ledgerAddress = toLedgerAddr $ shelleyAddressInEra @Api.Era Api.ShelleyBasedEraBabbage address
```
This is a bit of an odd piece of code. What we _could_ do here:
* use `mkVkAddress` to easily create an `AddressInEra` and see if we can avoid the `Address ShelleyAddr`, or
* stick with the `Address ShelleyAddr` as this actually does not need an `era` and can always be converted to a ledger `Addr`

Concretely for this PR, I would suggest to avoid mentioning babbage here by:
```suggestion
  ledgerAddress = toLedgerAddr $ shelleyAddressInEra @Api.Era Api.shelleyBasedEra address
```

ch1bo commented on hydra-node/src/Hydra/Chain/CardanoClient.hs:
```diff
@@ -76,6 +76,7 @@ buildTransaction networkId socket changeAddress utxoToSpend collateral outs = do
   pure $
     second balancedTxBody $
       makeTransactionBodyAutoBalance
+        ShelleyBasedEraBabbage
```
Should: use at least `shelleyBasedEra` to not mention `Babbage`

Could: shim `makeTransactionBodyAutoBalance` to not take this witness but produce it with `IsShelleyBasedEra(shelleyBasedEra)`

ch1bo replied:
```suggestion
        shelleyBasedEra
```

ch1bo commented on hydra-node/src/Hydra/Ledger/Cardano/Evaluate.hs:
```diff
@@ -222,7 +222,8 @@ prepareTxScripts tx utxo = do
     let pArgs = Ledger.getPlutusData <$> arguments
-    appliedTerm <- left show $ mkTermToEvaluate Plutus.PlutusV2 protocolVersion script pArgs
+    x <- left show $ Plutus.deserialiseScript Plutus.PlutusV2 protocolVersion script
+    appliedTerm <- left show $ mkTermToEvaluate Plutus.PlutusV2 protocolVersion x pArgs
```
:+1:

ch1bo commented on hydra-chain-observer/src/Hydra/ChainObserver.hs:
```diff
@@ -152,7 +153,7 @@ chainSyncClient tracer networkId startingPoint =
       { recvMsgRollForward = \blockInMode _tip -> ChainSyncClient $ do
           case blockInMode of
-            BlockInMode (Block _header txs) BabbageEraInCardanoMode -> do
+            BlockInMode BabbageEra (Block _header txs) BabbageEraInCardanoMode -> do
```
Seeing this one really makes you wonder _"why another witness?"_.

The change here is fine, but this really makes we sad and want to stop using `cardano-api` right away.

You _could_ just not match on it like you did in the `hydra-node`
```suggestion
            BlockInMode _ (Block _header txs) BabbageEraInCardanoMode -> do
```

ch1bo commented on hydra-cluster/src/Hydra/Cluster/Faucet.hs:
```diff
@@ -71,7 +71,7 @@ seedFromFaucet node@RunningNode{networkId, nodeSocket} receivingVerificationKey
   theOutput =
     TxOut
-      (shelleyAddressInEra receivingAddress)
+      (shelleyAddressInEra ShelleyBasedEraBabbage receivingAddress)
```
```suggestion
      (shelleyAddressInEra shelleyBasedEra receivingAddress)
```

See similar comment about weirdness of AddressInEra.

ch1bo commented on hydra-cluster/src/CardanoClient.hs:
```diff
@@ -50,6 +50,7 @@ calculateMinFee networkId body Sizes{inputs, outputs, witnesses} pparams =
   let tx = makeSignedTransaction [] body
       noByronWitnesses = 0
    in estimateTransactionFee
+        ShelleyBasedEraBabbage
```
```suggestion
        shelleyBasedEra
```

ch1bo commented on hydra-cluster/src/CardanoClient.hs:
```diff
@@ -40,7 +40,7 @@ buildScriptAddress script networkId =
 buildRaw ins outs fee =
   createAndValidateTransactionBody $
-    defaultTxBodyContent
+    defaultTxBodyContent BabbageEra
```
I would suggest to also shim `defaultTxBodyContent` like we did it for `createAndValidateTransactionBody`

ch1bo commented on hydra-cardano-api/src/Hydra/Cardano/Api/Tx.hs:
```diff
@@ -115,14 +111,18 @@ fromLedgerTx ledgerTx =
+      TxScriptValidity AlonzoEraOnwardsBabbage ScriptInvalid
+
+makeShelleyKeyWitness :: TxBody Era -> ShelleyWitnessSigningKey -> KeyWitness Era
+makeShelleyKeyWitness = Cardano.Api.makeShelleyKeyWitness ShelleyBasedEraBabbage
```
Should: move this to `Hydra.Cardano.Api` which is the intended home for this `Era`-specific functions and patterns (see README.md and module haddocks)

This way, you also do not need to qualify the `makeShelleyKeyWitness` because these sub-modules won't have the top-level `Api.hs` in scope.

ch1bo replied:
Done in https://github.com/input-output-hk/hydra/commit/67e685a529ba4aea9cefaecfe358ba8604a9634e

ch1bo reviewed (changes requested):
This is looking good now! Only a few suggestions and one commit to drop before we merge this.

Please update the PR title and provide a description _why_ this is needed / how it is related to some issue (I think it is needed, but PRs or commits should be documenting this)

Also, please honor the PR template and update at least the CHANGELOG.

ch1bo committed:
Submit NewTx transactions as CBOR in an envelope in benchmark

This ensures that the byte-for-byte decoded data set is ingested into a
Hydra head correctly. The JSON instance of transactions might not
preserve the body byte-for-byte and hence invalidate witnesses leading
to invalid transaction.

abailly committed:
Only extract txId from JSON-serialised Tx in benchmarks

Seems like we are having issues with inconsistent To/FromJSON
serialisation following upgrade, so perhaps we should move our logs to
use the "canonical" CBOR represenation of a Tx instead of semi-custom JSON?

ch1bo committed:
Rename MultiAssetSupportedInEra -> MaryEraOnwards

This was renamed upstream and we should capture the meaning the same
way. It is to provide witnesses for features which are only available
from mary era onwards.

ch1bo committed:
Rename HasInlineDatums -> BabbageEraOnwards

Same rationale as previous commit, got renamed upstream.

ch1bo committed:
Rename HasScriptData -> IsAlonzoEraOnwards

Same rationale as previous two commits

ch1bo committed:
Move (latest) Era-specific helpers to Hydra.Cardano.Api module

These re-definitions are to avoid additional witness passing as long as
we are just fine with the one specific / latest era (which we are by
using Hydra.Cardano.Api). Consequently and according to the package
README, these belong to the Api.hs module.

ch1bo committed:
Create an Era-specific wrapper for defaultTxBodyContent

This avoids mentioning "Babbage" in downstream packages like
hydra-cluster and will reduce the number of places to switch.

ch1bo committed:
Create an Era-specific wrapper for signShelleyTransaction

This avoids mentioning "Babbage" in downstream packages like
hydra-cluster and will reduce the number of places to switch.

ch1bo committed:
Do not mention Babbage outside of hydra-cardano-api

This reduces the number of times Babbage is mentioned in the hydra-node package and will help switching eras.

ch1bo reviewed (approved):
Addressed all my concerns and this can be merged now IMO

ch1bo committed:
Use recent cabal index-states

This removes "Requested index-state .. is newer than" warnings when
building with cabal.
